packages/astro/src/assets/fonts/vite-plugin-fonts.ts

```diff
@@ -0,0 +1,306 @@
+				};
+			},
+		};
+	}
+
+	const providers: Array<FontProvider<string>> = settings.config.experimental.fonts.providers ?? [];
+	const families: Array<FontFamily<string>> = settings.config.experimental.fonts.families;
+
+	// We don't need to take the trailing slash and build output configuration options
+	// into account because we only serve (dev) or write (build) static assets (equivalent
+	// to trailingSlash: never)
+	const baseUrl = removeTrailingForwardSlash(settings.config.base) + URL_PREFIX;
+
+	let resolvedMap: Map<string, { preloadData: PreloadData; css: string }> | null = null;
+	// Key is `${hash}.${ext}`, value is a URL.
+	// When a font file is requested (eg. /_astro/fonts/abc.woff), we use the hash
+	// to download the original file, or retrieve it from cache
+	let hashToUrlMap: Map<string, string> | null = null;
+	let isBuild: boolean;
+	let cache: Cache['cache'] | null = null;
+
+	async function initialize(moduleLoader?: ModuleLoader) {
+		const { h64ToString } = await xxhash();
+
+		const resolved = await resolveProviders({
+			settings,
+			providers,
+			moduleLoader,
+		});
+
+		const storage = createStorage({
+			// In dev, we cache fonts data in .astro so it can be easily inspected and cleared
+			base: isBuild
+				? new URL(CACHE_DIR, settings.config.cacheDir)
+				: new URL(CACHE_DIR, settings.dotAstroDir),
+		});
+		cache = createCache({ storage }).cache;
+
+		const { resolveFont } = await unifont.createUnifont(
+			resolved.map((e) => e.provider(e.config)),
+			{ storage },
+		);
+		// We initialize shared variables here and reset them in buildEnd
+		// to avoid locking memory
+		resolvedMap = new Map();
+		hashToUrlMap = new Map();
+		for (const family of families) {
+			const resolvedOptions: unifont.ResolveFontOptions = {
+				weights: family.weights ?? DEFAULTS.weights,
+				styles: family.styles ?? DEFAULTS.styles,
+				subsets: family.subsets ?? DEFAULTS.subsets,
+				fallbacks: family.fallbacks ?? DEFAULTS.fallbacks,
+			};
+			const { fonts: fontsData, fallbacks } = await resolveFont(family.name, resolvedOptions, [
+				family.provider,
+			]);
+
+			// TODO: investigate using fontaine for fallbacks
+			const preloadData: PreloadData = [];
+			let css = '';
+			for (const data of fontsData) {
+				for (const source of data.src as unknown as Array<Record<string, string>>) {
+					// Types are wonky but a source is
+					// 1. local and has a name
+					// 2. remote and has an url
+					// Once we have the key, it's safe to access the related source property
+					const key = 'name' in source ? 'name' : 'url';
+					const value = source[key];
+					const hash = h64ToString(value) + extname(value);
+					const url = baseUrl + hash;
+					if (!hashToUrlMap.has(hash)) {
+						hashToUrlMap.set(hash, value);
+						const segments = hash.split('.');
+						// It's safe, there's at least 1 member in the array
+						const type = segments.at(-1)!;
+						if (segments.length === 1 || !FONT_TYPES.includes(type)) {
```

Comment:
The comment on line 196 is in contradiction with `segments.length === 1`. The check will always be false because `segments` will always have two items: the hash and the extension/type. This is a bug?

Reply:
I tested manually this part and basically consider those cases (assuming a split on dots):
- `""` => `[""]` (length 1 invalid)
- `"abc"` => `["abc"]` (length 1 invalid)
- `"abc.def"` => `["abc", "def"]` (length 2 valid, last member not matching font type)
- `"abc.woff"` => `["abc", "woff"]` (length 2 valid, last member matching font type)
- `"abc.def.ghi"` => `["abc", "def", "ghi"]` (length 3 valid, last member not matching font type)
- `"abc.def.ttf"` => `["abc", "def", "ttf"]` (length 3 valid, last member matching font type)